components/sync15/src/state.rs

```diff
@@ -852,6 +853,189 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_from_previous_state_declined() {
+        let _ = env_logger::try_init();
+        // The state-machine sequence where we didn't use the previous state
+        // (ie, where the state machine restarted)
+        let sm_seq_restarted = vec![
+            "WithPreviousState",
+            "InitialWithConfig",
+            "InitialWithInfo",
+            "InitialWithMetaGlobal",
+            "Ready",
+        ];
+        // The state-machine sequence where we used the previous state.
+        let sm_seq_used_previous = vec!["WithPreviousState", "Ready"];
+
+        // do the actual test.
+        // (clippy gets upset with `expected_states: &Vec<&str>` - wants it
+        // declared as `&[&str]` - which `assert_eq!()` then fails on!)
+        #[allow(clippy::ptr_arg)]
+        fn do_test(
+            client: &dyn SetupStorageClient,
+            root_key: &KeyBundle,
+            mut pgs: &mut PersistedGlobalState,
+            engine_updates: Option<&HashMap<String, bool>>,
+            old_state: GlobalState,
+            expected_states: &Vec<&str>,
+        ) {
+            let mut state_machine = SetupStateMachine::for_full_sync(
+                client,
+                root_key,
+                &mut pgs,
+                engine_updates,
+                &NeverInterrupts,
+            );
+            assert!(
+                state_machine.run_to_ready(Some(old_state)).is_ok(),
+                "Should drive state machine to ready"
+            );
+            assert_eq!(state_machine.sequence, *expected_states);
+        }
+
+        // and all the complicated setup...
+        let ts_metaglobal = 123_456;
+        let ts_keys = 145_000;
+        let root_key = KeyBundle::new_random().unwrap();
+        let keys = CollectionKeys {
+            timestamp: ServerTimestamp(123_400),
```

Comment:
Where does this old timestamp come into play? Or do we always overwrite it with `ts_keys`?

Reply:
Thanks for noticing and it's a great question!

This is the timestamp we'll get back if we re-fetch the keys collection (which we do in most of the new tests). While the test doesn't care, I agree it's confusing, so I changed it to `ServerTimestamp(ts_keys + 1)` (which is hopefully enough of a hint)